File: Apps/Desktop/Sources/Threads/ProjectIsolator.swift
import Darwin
import Foundation

/// Copy-on-write project clones for isolated Threads.
enum ProjectIsolator {
    static var defaultRoot: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appending(path: ".fission/worktrees", directoryHint: .isDirectory)
    }

    static func create(
        from workingDirectory: String,
        isolateRoot: URL,
        requestedBranch: String? = nil,
        makeIdentifier: () -> String
    ) throws -> String {
        let selectedDirectory = URL(fileURLWithPath: workingDirectory).standardizedFileURL
        let sourceRoot = try resolvedSourceRoot(from: selectedDirectory)
        let relativePath = relativeSubpath(of: selectedDirectory, under: sourceRoot)
        let projectName = sourceRoot.lastPathComponent
        guard !projectName.isEmpty else {
            throw IsolateError.invalidProject
        }

        let repositoryIsolates = isolateRoot.appending(path: projectName, directoryHint: .isDirectory)
        let resolved = try resolveBranchAndIsolate(
            requestedBranch: requestedBranch,
            sourceRoot: sourceRoot,
            repositoryIsolates: repositoryIsolates,
            projectName: projectName,
            makeIdentifier: makeIdentifier
        )

        try FileManager.default.createDirectory(
            at: resolved.isolateDirectory,
            withIntermediateDirectories: true
        )
        do {
            try assertCopyOnWriteAvailable(from: sourceRoot, to: resolved.isolateDirectory)
            try cloneDirectory(from: sourceRoot, to: resolved.destination)
            try createBranchIfNeeded(at: resolved.destination, named: resolved.branch)
        } catch {
            try? FileManager.default.removeItem(at: resolved.isolateDirectory)
            throw error
        }

        guard !relativePath.isEmpty else { return resolved.destination.path }
        return resolved.destination.appending(path: relativePath).path
    }

    static func removeIsolate(workingDirectory: String?, isolateRoot: URL) throws {
        guard let isolateDirectory = isolateDirectory(
            for: workingDirectory,
            isolateRoot: isolateRoot
        ) else {
            return
        }
        try FileManager.default.removeItem(at: isolateDirectory)
    }

    static func isolateDirectory(for workingDirectory: String?, isolateRoot: URL) -> URL? {
        guard let workingDirectory else { return nil }
        let cwd = URL(fileURLWithPath: workingDirectory).standardizedFileURL
        let root = isolateRoot.standardizedFileURL
        let prefix = root.path.hasSuffix("/") ? root.path : root.path + "/"
        guard cwd.path.hasPrefix(prefix) else { return nil }

        let relative = String(cwd.path.dropFirst(prefix.count))
        let parts = relative.split(separator: "/").map(String.init)
        guard let project = parts.first,
              let leafIndex = parts.dropFirst().firstIndex(of: project) else {
            return nil
        }
        return parts[0..<leafIndex].reduce(root) { directory, component in
            directory.appending(path: component, directoryHint: .isDirectory)
        }
    }

    private static func resolveBranchAndIsolate(
        requestedBranch: String?,
        sourceRoot: URL,
        repositoryIsolates: URL,
        projectName: String,
        makeIdentifier: () -> String
    ) throws -> (branch: String, isolateDirectory: URL, destination: URL) {
        if let requestedBranch {
            guard GitWorktreeBranch.isValid(requestedBranch) else {
                throw IsolateError.invalidBranchName
            }
            let isolateDirectory = repositoryIsolates.appending(
                path: requestedBranch,
                directoryHint: .isDirectory
            )
            if FileManager.default.fileExists(atPath: isolateDirectory.path) {
                throw IsolateError.isolateExists(requestedBranch)
            }
            if sourceHasGitBranch(named: requestedBranch, at: sourceRoot) {
                throw IsolateError.branchAlreadyExists(requestedBranch)
            }
            return (
                requestedBranch,
                isolateDirectory,
                isolateDirectory.appending(path: projectName, directoryHint: .isDirectory)
            )
        }

        var branch = ""
        var isolateDirectory = repositoryIsolates
        repeat {
            branch = "fission-\(makeIdentifier())"
            isolateDirectory = repositoryIsolates
                .appending(path: branch, directoryHint: .isDirectory)
        } while FileManager.default.fileExists(atPath: isolateDirectory.path)
            || sourceHasGitBranch(named: branch, at: sourceRoot)
        return (
            branch,
            isolateDirectory,
            isolateDirectory.appending(path: projectName, directoryHint: .isDirectory)
        )
    }

    private static func resolvedSourceRoot(from selectedDirectory: URL) throws -> URL {
        if let repository = gitToplevel(at: selectedDirectory) {
            let git = repository.appending(path: ".git")
            var isDirectory: ObjCBool = false
            guard FileManager.default.fileExists(atPath: git.path, isDirectory: &isDirectory) else {
                throw IsolateError.invalidGitLayout
            }
            guard isDirectory.boolValue else {
                throw IsolateError.linkedWorktree
            }
            return repository
        }
        return selectedDirectory
    }

    private static func relativeSubpath(of selectedDirectory: URL, under sourceRoot: URL) -> String {
        selectedDirectory.path
            .dropFirst(sourceRoot.path.count)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }

    private static func gitToplevel(at directory: URL) -> URL? {
        guard let path = try? runGit([
            "-C", directory.path,
            "rev-parse", "--show-toplevel"
        ]) else {
            return nil
        }
        let trimmed = path.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        return URL(fileURLWithPath: trimmed).standardizedFileURL
    }

    private static func sourceHasGitBranch(named branch: String, at sourceRoot: URL) -> Bool {
        guard let listed = try? runGit([
            "-C", sourceRoot.path,
            "branch", "--list", "--", branch
        ]) else {
            return false
        }
        return !listed.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    private static func createBranchIfNeeded(at destination: URL, named branch: String) throws {
        let git = destination.appending(path: ".git")
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: git.path, isDirectory: &isDirectory),
              isDirectory.boolValue else {
            return
        }
        guard let head = try? runGit([
            "-C", destination.path,
            "rev-parse", "--verify", "HEAD"
        ]), !head.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return
        }
        _ = try runGit(["-C", destination.path, "switch", "-c", branch])
    }

    private static func assertCopyOnWriteAvailable(from source: URL, to destinationParent: URL) throws {
        guard isAPFS(source), isAPFS(destinationParent) else {
            throw IsolateError.copyOnWriteUnavailable
        }
        guard let sourceDevice = deviceID(of: source),
              let destinationDevice = deviceID(of: destinationParent),
              sourceDevice == destinationDevice else {
            throw IsolateError.copyOnWriteUnavailable
        }
        try probeClonefile(from: source, to: destinationParent)
    }

    private static func probeClonefile(from source: URL, to destinationParent: URL) throws {
        guard let probeSource = firstRegularFile(in: source) else { return }
        let probeDestination = destinationParent.appending(path: ".fission-cow-probe")
        try cloneFile(from: probeSource, to: probeDestination)
        try FileManager.default.removeItem(at: probeDestination)
    }

    private static func firstRegularFile(in directory: URL) -> URL? {
        guard let enumerator = FileManager.default.enumerator(
            at: directory,
            includingPropertiesForKeys: [.isRegularFileKey],
            options: [.skipsHiddenFiles]
        ) else {
            return nil
        }
        while let file = enumerator.nextObject() as? URL {
            let isRegular = (try? file.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile) ?? false
            if isRegular { return file }
        }
        return nil
    }

    private static func cloneDirectory(from source: URL, to destination: URL) throws {
        let status = source.path.withCString { sourcePath in
            destination.path.withCString { destinationPath in
                copyfile(
                    sourcePath,
                    destinationPath,
                    nil,
                    copyfile_flags_t(COPYFILE_CLONE | COPYFILE_RECURSIVE)
                )
            }
        }
        guard status == 0 else {
            throw IsolateError.cloneFailed(errno)
        }
    }

    private static func cloneFile(from source: URL, to destination: URL) throws {
        let status = source.path.withCString { sourcePath in
            destination.path.withCString { destinationPath in
                clonefile(sourcePath, destinationPath, 0)
            }
        }
        guard status == 0 else {
            throw IsolateError.copyOnWriteUnavailable
        }
    }

    private static func isAPFS(_ url: URL) -> Bool {
        var filesystem = statfs()
        guard statfs(url.path, &filesystem) == 0 else { return false }
        return withUnsafeBytes(of: filesystem.f_fstypename) { buffer in
            guard let base = buffer.baseAddress?.assumingMemoryBound(to: CChar.self) else {
                return false
            }
            return String(cString: base) == "apfs"
        }
    }

    private static func deviceID(of url: URL) -> dev_t? {
        var info = stat()
        guard lstat(url.path, &info) == 0 else { return nil }
        return info.st_dev
    }

    private static func runGit(_ arguments: [String]) throws -> String {
        let process = Process()
        let output = Pipe()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/git")
        process.arguments = arguments
        process.standardOutput = output
        process.standardError = output

        do {
            try process.run()
        } catch {
            throw IsolateError.gitUnavailable
        }

        let data = output.fileHandleForReading.readDataToEndOfFile()
        process.waitUntilExit()
        let message = (String(bytes: data, encoding: .utf8) ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard process.terminationStatus == 0 else {
            throw IsolateError.commandFailed(message)
        }
        return message
    }
}

enum IsolateError: LocalizedError {
    case invalidProject
    case invalidGitLayout
    case linkedWorktree
    case invalidBranchName
    case branchAlreadyExists(String)
    case isolateExists(String)
    case copyOnWriteUnavailable
    case cloneFailed(Int32)
    case gitUnavailable
    case commandFailed(String)

    var errorDescription: String? {
        switch self {
        case .invalidProject:
            "The selected project folder is invalid."
        case .invalidGitLayout:
            "The selected Git repository could not be read."
        case .linkedWorktree:
            "Linked Git worktrees cannot be isolated. Choose the main repository folder."
        case .invalidBranchName:
            "Enter a valid Git branch name."
        case let .branchAlreadyExists(branch):
            "A branch named \"\(branch)\" already exists."
        case let .isolateExists(branch):
            "An isolated workspace for \"\(branch)\" already exists."
        case .copyOnWriteUnavailable:
            "The project must be on APFS on the same volume as ~/.fission/worktrees."
        case .cloneFailed:
            "The project folder could not be copied."
        case .gitUnavailable:
            "Git could not be started."
        case let .commandFailed(message):
            message.isEmpty ? "Git could not prepare the isolated workspace." : message
        }
    }
}

File: packages/electron/src/shared/projectIsolator.ts
import { existsSync, mkdirSync, rmSync, statSync, cpSync, constants } from "node:fs";
import { homedir } from "node:os";
import { basename, dirname, join, relative, resolve } from "node:path";
import { execFileSync } from "node:child_process";
import * as gitWorktreeBranch from "./gitWorktreeBranch";

export class IsolateError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "IsolateError";
  }
}

export function defaultIsolateRoot(): string {
  return join(homedir(), ".fission", "worktrees");
}

export function createIsolate(input: {
  workingDirectory: string;
  isolateRoot?: string;
  requestedBranch?: string | null;
  makeIdentifier?: () => string;
}): string {
  const isolateRoot = resolve(input.isolateRoot ?? defaultIsolateRoot());
  const selectedDirectory = resolve(input.workingDirectory);
  const sourceRoot = resolvedSourceRoot(selectedDirectory);
  const relativePath = relativeSubpath(selectedDirectory, sourceRoot);
  const projectName = basename(sourceRoot);
  if (projectName.length === 0) {
    throw new IsolateError("The selected project folder is invalid.");
  }

  const repositoryIsolates = join(isolateRoot, projectName);
  const resolved = resolveBranchAndIsolate({
    requestedBranch: input.requestedBranch ?? null,
    sourceRoot,
    repositoryIsolates,
    projectName,
    makeIdentifier: input.makeIdentifier ?? randomIdentifier
  });

  mkdirSync(resolved.isolateDirectory, { recursive: true });
  try {
    awaitTestDelayIfRequested();
    cloneDirectory(sourceRoot, resolved.destination);
    createBranchIfNeeded(resolved.destination, resolved.branch);
  } catch (error) {
    rmSync(resolved.isolateDirectory, { recursive: true, force: true });
    throw error;
  }

  if (relativePath.length === 0) {
    return resolved.destination;
  }
  return join(resolved.destination, relativePath);
}

export function removeIsolate(workingDirectory: string | null, isolateRoot?: string): void {
  const isolateDirectory = isolateDirectoryFor(
    workingDirectory,
    resolve(isolateRoot ?? defaultIsolateRoot())
  );
  if (!isolateDirectory) {
    return;
  }
  rmSync(isolateDirectory, { recursive: true, force: true });
}

export function isolateDirectoryFor(
  workingDirectory: string | null | undefined,
  isolateRoot: string
): string | null {
  if (!workingDirectory) {
    return null;
  }
  const cwd = resolve(workingDirectory);
  const root = resolve(isolateRoot);
  const prefix = root.endsWith("/") ? root : `${root}/`;
  if (!cwd.startsWith(prefix)) {
    return null;
  }
  const parts = cwd.slice(prefix.length).split("/").filter(Boolean);
  const project = parts[0];
  if (!project) {
    return null;
  }
  const leafIndex = parts.findIndex((part, index) => index > 0 && part === project);
  if (leafIndex === -1) {
    return null;
  }
  return join(root, ...parts.slice(0, leafIndex));
}

function resolveBranchAndIsolate(input: {
  requestedBranch: string | null;
  sourceRoot: string;
  repositoryIsolates: string;
  projectName: string;
  makeIdentifier: () => string;
}): { branch: string; isolateDirectory: string; destination: string } {
  if (input.requestedBranch) {
    if (!gitWorktreeBranch.isValid(input.requestedBranch)) {
      throw new IsolateError("Enter a valid Git branch name.");
    }
    const isolateDirectory = join(input.repositoryIsolates, input.requestedBranch);
    if (existsSync(isolateDirectory)) {
      throw new IsolateError(
        `An isolated workspace for "${input.requestedBranch}" already exists.`
      );
    }
    if (sourceHasGitBranch(input.requestedBranch, input.sourceRoot)) {
      throw new IsolateError(`A branch named "${input.requestedBranch}" already exists.`);
    }
    return {
      branch: input.requestedBranch,
      isolateDirectory,
      destination: join(isolateDirectory, input.projectName)
    };
  }

  let branch = "";
  let isolateDirectory = input.repositoryIsolates;
  do {
    branch = `fission-${input.makeIdentifier()}`;
    isolateDirectory = join(input.repositoryIsolates, branch);
  } while (existsSync(isolateDirectory) || sourceHasGitBranch(branch, input.sourceRoot));

  return {
    branch,
    isolateDirectory,
    destination: join(isolateDirectory, input.projectName)
  };
}

function resolvedSourceRoot(selectedDirectory: string): string {
  const repository = gitToplevel(selectedDirectory);
  if (!repository) {
    return selectedDirectory;
  }
  const git = join(repository, ".git");
  if (!existsSync(git)) {
    throw new IsolateError("The selected Git repository could not be read.");
  }
  if (!statSync(git).isDirectory()) {
    throw new IsolateError(
      "Linked Git worktrees cannot be isolated. Choose the main repository folder."
    );
  }
  return repository;
}

function relativeSubpath(selectedDirectory: string, sourceRoot: string): string {
  const value = relative(sourceRoot, selectedDirectory);
  return value === "." ? "" : value;
}

function gitToplevel(directory: string): string | null {
  try {
    return runGit(["-C", directory, "rev-parse", "--show-toplevel"]);
  } catch {
    return null;
  }
}

function sourceHasGitBranch(branch: string, sourceRoot: string): boolean {
  try {
    const listed = runGit(["-C", sourceRoot, "branch", "--list", "--", branch]);
    return listed.length > 0;
  } catch {
    return false;
  }
}

function createBranchIfNeeded(destination: string, branch: string): void {
  const git = join(destination, ".git");
  if (!existsSync(git) || !statSync(git).isDirectory()) {
    return;
  }
  try {
    const head = runGit(["-C", destination, "rev-parse", "--verify", "HEAD"]);
    if (head.length === 0) {
      return;
    }
    runGit(["-C", destination, "switch", "-c", branch]);
  } catch {
    return;
  }
}

function cloneDirectory(source: string, destination: string): void {
  mkdirSync(dirname(destination), { recursive: true });
  try {
    if (process.platform === "darwin") {
      assertCopyOnWriteAvailable(source, dirname(destination));
      try {
        cpSync(source, destination, {
          recursive: true,
          verbatimSymlinks: true,
          mode: copyOnWriteMode()
        });
        return;
      } catch (error) {
        if (error instanceof IsolateError) {
          throw error;
        }
        throw new IsolateError(
          "The project must be on APFS on the same volume as ~/.fission/worktrees."
        );
      }
    }
    cpSync(source, destination, { recursive: true, verbatimSymlinks: true });
  } catch (error) {
    if (error instanceof IsolateError) {
      throw error;
    }
    throw new IsolateError("The project folder could not be copied.");
  }
}

function copyOnWriteMode(): number {
  const flags = constants as unknown as Record<string, number | undefined>;
  return flags.COPYFILE_FICLONE_FORCE ?? flags.COPYFILE_FICLONE ?? 0;
}

function assertCopyOnWriteAvailable(source: string, destinationParent: string): void {
  try {
    const sourceDevice = statSync(source).dev;
    mkdirSync(destinationParent, { recursive: true });
    const destinationDevice = statSync(destinationParent).dev;
    if (sourceDevice !== destinationDevice) {
      throw new IsolateError(
        "The project must be on APFS on the same volume as ~/.fission/worktrees."
      );
    }
  } catch (error) {
    if (error instanceof IsolateError) {
      throw error;
    }
  }
}

function awaitTestDelayIfRequested(): void {
  const raw = process.env.FISSION_ISOLATE_DELAY_MS;
  if (!raw) {
    return;
  }
  const milliseconds = Number.parseInt(raw, 10);
  if (!Number.isFinite(milliseconds) || milliseconds <= 0) {
    return;
  }
  const capped = Math.min(milliseconds, 10_000);
  Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, capped);
}

function randomIdentifier(): string {
  const characters = "abcdefghijklmnopqrstuvwxyz0123456789";
  let result = "";
  for (let index = 0; index < 6; index += 1) {
    result += characters[Math.floor(Math.random() * characters.length)];
  }
  return result;
}

function runGit(arguments_: string[]): string {
  try {
    return execFileSync("git", arguments_, { encoding: "utf8" }).trim();
  } catch {
    throw new IsolateError("Git could not be started.");
  }
}
